src/gosling-track/data-abstraction.ts

```diff
@@ -199,6 +201,68 @@ export function getTabularData(
                 }
             });
         });
+    } else if (spec.data.type === 'matrix') {
+        if (!data.dense || typeof data.tileY === 'undefined' || typeof data.tileHeight === 'undefined') {
+            // we do not have sufficient data.
+            return;
+        }
+
+        // width and height of the tile
+        const numBins = Math.sqrt(data.dense.length);
+
+        const { tileX, tileY, tileWidth, tileHeight } = data;
+        const numericValues = data.dense;
+
+        const tileXUnitSize = tileWidth / numBins;
+        const tileYUnitSize = tileHeight / numBins;
+        const columnField = spec.data.column ?? 'x';
+        const rowField = spec.data.row ?? 'y';
+
+        // TODO: a way to improve rendering performance?
+        // For the rendering performance, we aggregate cells so that we draw smaller number of marks.
+        const aggSize = 16; // assuming that # of cells can be divided by binSize, which is mostly 256 or 1024
+        for (let i = 0; i < numericValues.length / aggSize; i++) {
+            const aggLen = Math.sqrt(aggSize);
+            const xIndex = (i * aggLen) % numBins;
+            const yIndex = Math.floor((i * aggLen) / numBins) * aggLen;
+
+            // Being xIndex and yIndex the top-let origin, aggregate 4 x 4 cells
+            let value = NaN;
+            for (let c = 0; c < aggLen; c++) {
+                for (let r = 0; r < aggLen; r++) {
+                    const curVal = numericValues[(yIndex + r) * numBins + (xIndex + c)];
+                    if (!isNaN(+curVal)) {
+                        if (isNaN(value)) {
+                            value = 0;
+                        }
+                        value += curVal;
+                    }
+                }
+            }
+
+            if (isNaN(value)) {
+                // this is a missing value
+                continue;
+            }
+
+            value /= aggSize;
+
+            const xs = tileX + xIndex * tileXUnitSize;
+            const xe = tileX + (xIndex + aggLen) * tileXUnitSize;
+            const ys = tileY + yIndex * tileYUnitSize;
+            const ye = tileY + (yIndex + aggLen) * tileYUnitSize;
+            const x = (xs + xe) / 2.0;
+            const y = (ys + ye) / 2.0;
+            tabularData.push({
+                value,
+                [columnField]: x,
+                [`${columnField}s`]: xs,
+                [`${columnField}e`]: xe,
+                [rowField]: y,
+                [`${rowField}s`]: ys,
+                [`${rowField}e`]: ye
+            });
```

Comment:
This is a great question. There is certainly a strong tradeoff between flexible encoding and efficient rendering.

In this PR, I went with flexibility since the core feature of our grammar is to support expressiveness. And to support expressive encoding, some kinds of data abstraction (here, converting to columnar formats) is necessary. In this way, one can fully use the concept of our grammar, like using other mark types (e.g., point w/ varying size and color), data transformation (e.g., filter data points with very strong signals), or applying semantic zooming (e.g., showing text labels when zoomed in very far).

I think there is some space to improve the performance (especially on the rendering pipeline), but I don't currently have a concrete idea on how to make it happen, both in the data abstraction and rendering, without sacrificing the flexibility. I would need to test this in-depth. One thing I would need to refer to is the rendering pipeline of GenomeSpy since it looks to work very smoothly with columnar data (CSV files).

Additionaly, I think, we can support the fast HiGlass-based matrix rendering when a user use a certain `template` (e.g., `{ template: "matrix", data: { /* Cooler data */ } }`). Since this is a template, we do not need to worry about whether a user change encoding that is not defined by the template (e.g., changing rect to point). We will be able to allow overlaying annotations (e.g., `alignment: "overlay", tracks: [ { /* matrix template */ }, { /* annotation rules */ }]`).

Reply:
To add, my main idea on this PR was to perform the data abstraction for the expressive encoding (since it is necessary for the grammar), but improve the performance (especially on the rendering) as a follow-up PRs, to make the zooming/panning smooth. When I tested locally and informally, the rendering pipeline seemed to be the main bottleneck.

Let me know if it makes more sense to make this PR pending until we improve the performance, or you have any other better ideas to work around this issue.

Reply:
Thanks for the detailed response! I agree (mostly) with the prioritization of flexibility over efficiency in this regard. I guess my point, specifically, is questioning why we map the dense matrix data to `tabularData`. 

I understand why having this mapping is important for columnar data (that you want to map to track with `x` === `genomic` and `y` === `genomic`, but I'm struggling to see what flexibility this provides in the context of dense matrix data. 
 
I guess now you could do something like,

```javascript
{
     data: {
          url: GOSLING_PUBLIC_DATA.matrixMicroC,
          type: 'matrix'
      },
      mark: 'point',
      x: { field: 'xs', type: 'genomic', axis: 'none' },
      y: { field: 'ys', type: 'genomic', axis: 'none' },
      color: { field: 'value', type: 'quantitative', range: 'warm' },
}
```

but it isn't clear to me how this is useful. 

I bet I'm conceptually missing the vision of this PR, but I think ideally we should make the mapping of matrix -> tabular data "lazy" (e.g. only perform this conversion if we need to render something other than a standard heatmap). For example, if the configuration below is supplied, then we should just let HiGlass load the tiles natively,

```
{
     data: {
          url: GOSLING_PUBLIC_DATA.matrixMicroC,
          type: 'matrix'
      },
      mark: 'bar',
      x: { field: 'xs', type: 'genomic', axis: 'none' },
      xe: { field: 'xe', type: 'genomic', axis: 'none' },
      y: { field: 'ys', type: 'genomic', axis: 'none' },
      ye: { field: 'ye', type: 'genomic', axis: 'none' },
      color: { field: 'value', type: 'quantitative', range: 'warm' },
      width: 600,
      height: 600
}
```

I think to your point, it's important to remember that we can make the grammar expressive (allowing a wide variety of visualizations), but inherently _some_ of the possible visualizations will be more inefficient to render from others. In the implementation of the renderer, however, I think we should be careful to consider/optimize for common visualizations if possible.

Reply:
@manzt Thanks for the reply. Having a conversation on this issue is really helpful! 

The vision of this PR is to fully support the grammar approach. The encoding supported in HiGlass `matrix` is a very small space in the entire design space in our grammar (i.e., rectangular shape mark + quantitative color + two genomic axes), and we will need to figure out how to support the remaining spaces. For example, if one just adds stroke colors of marks or adjusts their sizes, we will not be able to use HiGlass due to its limited rendering expressiveness. I think there are many other encodings that users can try with using our grammar other than the regular matrix. One could try displaying arcs with colors in 1D tracks, focusing on showing strong interactions [^1], a bubble matrix [^2], or the superposition of two matrices with varying sizes for pair-wise comparison [^3]. Supporting semantic zooming will be another example (e.g., when dense, show the regular matrix representation, but when sparse, change representations and show details with labels).

Having said that, I agree with you that the regular matrix visualization in HiGlass will be the common encoding with the `matrix` data, and we will need to handle this nicely. I think, instead of using `tabularData` all the time, we can use the HiGlass rendering when a user uses the following spec. Until we find an efficient way to render the matrix directly on Gosling (e.g., using the lazy data processing), we can keep this logic in our grammar.

```
{
     data: {
          url: GOSLING_PUBLIC_DATA.matrixMicroC,
          type: 'matrix'
      },
      mark: 'bar',
      x: { field: 'xs', type: 'genomic', ... },
      xe: { field: 'xe', type: 'genomic', ... },
      y: { field: 'ys', type: 'genomic', ... },
      ye: { field: 'ye', type: 'genomic', ... },
      color: { field: 'value', type: 'quantitative' },
      width: ...,
      height: ...
}
```

To unify the HiGlass matrix with our grammar, I will need to test several things, such as overlaying gosling tracks on top of the higlass matrix (e.g., vertical and horizontal annotations) or using consistent color scales in the HiGlass matrix. Regarding lazy data processing, let me think about it further. At the moment, it is not clear to me (at least in the technical aspect) how we can best support flexible encoding using non-columnar data.

I can make changes on this PR to use the conditional HiGlass rendering, and I can test other issues in the follow-up PRs. What do you think?


[^1]: ![image](https://user-images.githubusercontent.com/9922882/146785620-1dc9cc83-1c81-468b-8f57-4af868d7d233.png)
[^2]: ![Screen Shot 2021-12-20 at 9 40 09 AM](https://user-images.githubusercontent.com/9922882/146784518-1966d10a-db39-4efd-9ab7-408b92f187fc.png)
[^3]: ![Screen Shot 2021-12-20 at 9 43 31 AM](https://user-images.githubusercontent.com/9922882/146785007-5230dc04-2151-4816-b936-f556b6a94869.png)

Reply:
> Having a conversation on this issue is really helpful!

Agreed!

> The vision of this PR is to fully support the grammar approach. 

I understand this now, and thank you for all the motivating examples. This will be great to have for future reference.

> I think, instead of using `tabularData` all the time, we can use the HiGlass rendering when a user uses the following spec. 

I think the idea of a conditional data transformation (depending on the spec) is a great compromise. This will be hidden from end users, offering flexibility in the grammar and efficiency when the data don't require transformation.

> I can make changes on this PR to use the conditional HiGlass rendering, and I can test other issues in the follow-up PRs. What do you think?

I'm happy to have this merged, but this sounds like a great plan in a follow up PR. Preferably this can be explored sooner than later, I think.